### PR TITLE
Adding prepacking to QLinearMatMul

### PR DESCRIPTION
**Pull request by chenfucn**
Add prepacking to QLinearMatMul.

### CODE REVIEW

File: onnxruntime/core/providers/cpu/math/quantize_linear_matmul.cc [hunk 1]
```diff
@@ -84,13 +86,22 @@ Status QLinearMatMul::Compute(OpKernelContext* ctx) const {
...
   gemm_params.ldc = gemm_params.N;
 
   for (size_t i = 0; i < helper.OutputOffsets().size(); i++) {
     gemm_params.A = a->template Data<uint8_t>() + helper.LeftOffsets()[i];
-    gemm_params.B = static_cast<const uint8_t*>(b->DataRaw()) + helper.RightOffsets()[i];
+    if (packed_b_) {
+      gemm_params.B = packed_b_.get();
+      gemm_params.BIsPacked = true;
+      gemm_params.BIsSigned = b_is_signed_;
+    } else if (b != nullptr) {
```
Comment (yufenglee) [inline onnxruntime/core/providers/cpu/math/quantize_linear_matmul.cc:98]: >(b != nullptr) { [](start = 14, length = 16)

nit: usually we don't check input tensor and assume it exists.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Comment by chenfucn** [comment]
Can't merge yet!

found a bug that missed by all tests. trying to fix it. I may need help on how to improve tests to catch similar bug.

File: onnxruntime/core/providers/cpu/math/matmul_integer.cc [hunk 2]
```diff
@@ -29,27 +41,43 @@ ONNX_OPERATOR_TYPED_KERNEL_EX(
...
-  ORT_RETURN_IF_ERROR(helper.Compute(a->Shape(), b ? b->Shape() : b_shape_));
-  Tensor* y = ctx->Output(0, helper.OutputShape());
+  const auto* a = ctx->Input<Tensor>(InIdx::A);
 
+  const uint8_t* b_data;
+  bool b_signed;
+  if (packed_b_) {
+    ORT_RETURN_IF_ERROR(helper.Compute(a->Shape(), b_shape_));
+    b_data = static_cast<const uint8_t*>(packed_b_.get());
+    b_signed = b_is_signed_;
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/math/matmul_integer.cc:52]: the preferred style would be "b_is_signed"

File: onnxruntime/core/providers/cpu/math/quantize_linear_matmul.cc [hunk 3]
```diff
@@ -30,21 +25,37 @@ ONNX_OPERATOR_KERNEL_EX(
...
-  const auto* a = ctx->Input<Tensor>(0);
-  const auto* b = ctx->Input<Tensor>(3);
-
   MatMulComputeHelper helper;
-  ORT_RETURN_IF_ERROR(helper.Compute(a->Shape(), b->Shape()));
-  Tensor* y = ctx->Output(0, helper.OutputShape());
+  const auto* a = ctx->Input<Tensor>(IN_A);
+
+  const uint8_t* b_start;
+  bool b_signed;  // can't modify b_is_signed_, this is a const method
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/math/quantize_linear_matmul.cc:32]: b_is_signed

File: onnxruntime/core/providers/cpu/math/quantize_linear_matmul.cc [hunk 4]
```diff
@@ -53,9 +64,9 @@ Status QLinearMatMul::Compute(OpKernelContext* ctx) const {
               "QLinearMatmul : result zero point must be a scalar or 1D tensor of size 1");
 
   // validate scale
-  const auto* a_scale = ctx->Input<Tensor>(1);
-  const auto* b_scale = ctx->Input<Tensor>(4);
-  const auto* y_scale = ctx->Input<Tensor>(6);
+  const auto* a_scale = ctx->Input<Tensor>(IN_Ascale);
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/math/quantize_linear_matmul.cc:67]: why does this use In_Ascale while the other classes use InIdx::Ascale

File: onnxruntime/test/providers/cpu/math/quantize_linear_matmul_test.cc [hunk 5]
```diff
@@ -192,5 +194,96 @@ TEST(QuantizeLinearMatmulOpTest, QLinearMatMulAllInputExceptT1AreInitializers) {
   QLinearMatMul2DTest(true);
 }
 
+/**
+ * @brief Extend QLinearMatMul for verifing prepacking behavior
```
Comment (tracysh) [inline onnxruntime/test/providers/cpu/math/quantize_linear_matmul_test.cc:198]: verifing->verifying

File: onnxruntime/core/providers/cpu/math/quantize_linear_matmul.cc (hunk 4 above)
Comment (chenfucn) [reply onnxruntime/core/providers/cpu/math/quantize_linear_matmul.cc]: changing all occurrences to be consistent  with  existing ones in dml

**Review by chenfucn — COMMENTED** [review]
Thanks Tracy. Fixed styling problems

**Review by tracysh — APPROVED** [review]

**Pull request merged**